File: packages/python/qsu/misc/retry.py
import time


def retry(func, options=None, **kwargs):
	opts = {**(options or {}), **kwargs}
	times = opts.get('times', 3)
	delay = opts.get('delay', 0)
	backoff = opts.get('backoff', 1)

	if times < 1:
		raise ValueError('`times` must be at least 1.')

	currentDelay = delay
	lastError = None

	for attempt in range(1, times + 1):
		try:
			return func()
		# `BaseException` is deliberately not caught, so `KeyboardInterrupt` and
		# `SystemExit` still stop the loop.
		except Exception as error:
			lastError = error

			# The delay sits *between* attempts, so the final failure is reported without
			# waiting one more time for nothing.
			if attempt < times:
				if currentDelay > 0:
					time.sleep(currentDelay / 1000)

				currentDelay *= backoff

	raise lastError

File: packages/javascript/lib/misc/retry.ts
import type { RetryOptions } from '../_types/global.js';
import { sleep } from './sleep.js';

export async function retry<T>(func: () => T | Promise<T>, options?: RetryOptions): Promise<T> {
	const times = options?.times ?? 3;
	const delay = options?.delay ?? 0;
	const backoff = options?.backoff ?? 1;

	if (times < 1) {
		throw new Error('`times` must be at least 1.');
	}

	let currentDelay = delay;
	let lastError: unknown;

	for (let attempt = 1; attempt <= times; attempt += 1) {
		try {
			return await func();
		} catch (error) {
			lastError = error;

			// The delay sits *between* attempts, so the final failure is reported without
			// waiting one more time for nothing.
			if (attempt < times) {
				if (currentDelay > 0) {
					await sleep(currentDelay);
				}

				currentDelay *= backoff;
			}
		}
	}

	throw lastError;
}
